tests/init/pos/instance-of-always-valid.scala

```diff
@@ -0,0 +1,13 @@
+final class MyAsInstanceOfClass(o: MyAsInstanceOfClass) {
+  val other: MyAsInstanceOfClass = {
+    if (o.asInstanceOf[Object] ne null) o // o is cold, but ne and AsInstanceOf is always valid
```

Comment:
It would be good to have tests to make sure that `o.asInstanceOf[Object]` has the right value. Specifically, a negative test for `o.asInstanceOf[Object].field` (would have to add a field to the class) and a positive test for `this.asInstanceOf[MyAsInstanceOfClass].field` (assuming the field comes before this expression).